pyemu/utils/helpers.py

```diff
@@ -90,6 +90,148 @@ def _pattern(self, pData):
+    passed_names = []
+    nz_names = pst.nnz_obs_names
+    [passed_names.extend(obs) for gs,obs in struct_dict.items()]
+    missing = list(set(passed_names) - set(nz_names))
+    if len(missing) > 0:
+        raise Exception("the following obs in struct_dict were not found in the nz obs names"+str(missing))
+    time_cols = ["time","datetime","distance"]
+    obs = pst.observation_data
+    if time_distance_col not in obs.columns:
+        raise Exception("time_distance_col missing")
+    dvals = obs.loc[passed_names,time_distance_col]
+    pobs = obs.loc[passed_names,:]
+    isna = pobs.loc[pd.isna(dvals),"obsnme"]
+    if isna.shape[0] > 0:
+        raise Exception("the following struct dict observations have NaN for time_distance_col: {0}".format(str(isna)))
+    if verbose:
+        print("--> getting full diagonal cov matrix")
+    fcov = pyemu.Cov.from_observation_data(pst)
+    fcov_dict = {o:np.sqrt(fcov.x[i]) for i,o in enumerate(fcov.names)}
+    if verbose:
+        print("-->draw full obs en from diagonal cov")
+    full_oe = pyemu.ObservationEnsemble.from_gaussian_draw(pst,fcov,num_reals=num_reals,fill=True)
+    for gs,onames in struct_dict.items():
+        if verbose:
+            print("-->processing cov matrix for {0} items with gs {1}".format(len(onames),gs))
+        dvals = obs.loc[onames,time_distance_col].values
+        gcov = gs.covariance_matrix(dvals,np.ones(len(onames)),names=onames)
+        if verbose:
+            print("...scaling rows and cols")
+        for i,name in enumerate(gcov.names):
+            gcov.x[:,i] *= fcov_dict[name]
+            gcov.x[i, :] *= fcov_dict[name]
+        if verbose:
+            print("...draw")
+        oe = pyemu.ObservationEnsemble.from_gaussian_draw(pst,gcov,num_reals=num_reals,fill=True,by_groups=False)
+        oe = oe.loc[:,gcov.names]
+        full_oe.loc[:,gcov.names] = oe._df.values
+
+    if enforce_bounds:
+        if verbose:
+            print("-->enforcing bounds")
+        ub_dict = {o:1e300 for o in full_oe.columns}
+        if "upper_bound" in pst.observation_data.columns:
+            ub_dict.update(pst.observation_data.upper_bound.fillna(1.0e300).to_dict())
+
+        lb_dict = {o:-1e300 for o in full_oe.columns}
+        if "lower_bound" in pst.observation_data.columns:
+            lb_dict.update(pst.observation_data.lower_bound.fillna(-1.0e200).to_dict())
+        allvals = full_oe.values
+        for i,name in enumerate(full_oe.columns):
+            #print("before:",name,ub_dict[name],full_oe.loc[:,name].max(),lb_dict[name],full_oe.loc[:,name].min())
+            #vals = full_oe.loc[:,name].values
+            vals = allvals[:,i]
+            vals[vals>ub_dict[name]] = ub_dict[name]
+            vals[vals < lb_dict[name]] = lb_dict[name]
+            #full_oe.loc[:,name] = vals#oe.loc[:,name].apply(lambda x: min(x,ub_dict[name])).apply(lambda x: max(x,lb_dict[name]))
+            allvals[:,i] = vals
```

Comment:
do these ever get passed back to full_oe?

Reply:
I checked this and the ref to `full_oe.values` remained intact - the testing covers this